contrib/babelfishpg_unit/test_ssl_read.c

```diff
@@ -0,0 +1,179 @@
+#include "babelfishpg_unit.h"
+#include "../babelfishpg_tds/src/include/tds_secure.h"
+
+#ifdef USE_SSL
+static char *prelogin_request;
+static int ReadPointer = 0;
+
+
+static ssize_t
+mock_socket_read(Port *port, void *ptr, size_t len)
+{
+
+    /*
+     *  Mock function of tds_secure_raw_read() present in tdssecure.c file
+     */
+
+    int i;
+    for (i = ReadPointer; i < ReadPointer + len; i++) 
+        sscanf(&prelogin_request[i * 2], "%2hhx", (unsigned char *)ptr + i);
+    ReadPointer += len;
+    return len;
+}
+
+
+TestResult*
+test_ssl_handshakeRead(void)
+{
+
+    /*
+     *  We will generate a prelogin request message and pass it to the handshake read function for processing
+     *  By comparing the number of bytes read against the expected value, we can verify the accuracy of the read operation 
+     */
+
+    BIO *h = NULL;
+    char *buf = NULL;
+    char *expected_str;
+    int expected;
+    int obtained;
+    size_t size_buf;
+    size_t size_binaryData;
+
+    TestResult* testResult = palloc0(sizeof(TestResult));
+    testResult->result = true;
+
+    h = BIO_new(BIO_s_mem());
+    ReadPointer = 0;
+
+    prelogin_request = strdup("1201000F0000010011A25E4571");
+    buf = malloc(strlen(prelogin_request)/2);
+    expected = strlen(prelogin_request)/2 - 8;
+
+    expected_str = (char *)malloc(strlen(prelogin_request)/2);
+    for (int i = 0; i < expected; i++) 
+        sscanf(&prelogin_request [i * 2], "%2hhx", (char *)expected_str + i);
+    
+    obtained = test_ssl_handshake_read(h, buf, expected, mock_socket_read, ReadPointer);
+
+    size_buf = strlen(buf);
+    size_binaryData = strlen(expected_str);
+    buf[size_buf+1] = '\0';
+    expected_str[size_binaryData+1] = '\0';
+    
+    TEST_ASSERT_TESTCASE(obtained == expected, testResult);
+    TEST_ASSERT_TESTCASE(*((unsigned char*)expected_str) == *((unsigned char*)buf), testResult);
```

Comment:
nit: please add comment saying which ASSERT is checking what? I mean add comment saying this ASSERT checks the number of bytes, this ASSERT checks the actual byte message. Also, as we are checking the number of bytes, I think we should check the actual message only iff the number of bytes matches or else there is no need to read. So, better to put and if condition maybe.

Reply:
Done in the next revision.